aws/resource_aws_service_discovery_public_dns_namespace.go

```diff
@@ -0,0 +1,144 @@
+package aws
+
+import (
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/servicediscovery"
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func resourceAwsServiceDiscoveryPublicDnsNamespace() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsServiceDiscoveryPublicDnsNamespaceCreate,
+		Read:   resourceAwsServiceDiscoveryPublicDnsNamespaceRead,
+		Delete: resourceAwsServiceDiscoveryPublicDnsNamespaceDelete,
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"description": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+			"arn": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"hosted_zone": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceAwsServiceDiscoveryPublicDnsNamespaceCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).sdconn
+
+	input := &servicediscovery.CreatePublicDnsNamespaceInput{
+		Name: aws.String(d.Get("name").(string)),
+	}
+
+	if v, ok := d.GetOk("description"); ok {
+		input.Description = aws.String(v.(string))
+	}
+
+	resp, err := conn.CreatePublicDnsNamespace(input)
+	if err != nil {
+		return err
+	}
+
+	stateConf := &resource.StateChangeConf{
+		Pending:    []string{servicediscovery.OperationStatusSubmitted, servicediscovery.OperationStatusPending},
+		Target:     []string{servicediscovery.OperationStatusSuccess},
+		Refresh:    servicediscoveryOperationRefreshStatusFunc(conn, *resp.OperationId),
+		Timeout:    5 * time.Minute,
+		Delay:      10 * time.Second,
+		MinTimeout: 3 * time.Second,
+	}
+
+	opresp, err := stateConf.WaitForState()
+	if err != nil {
+		return err
+	}
+
+	d.SetId(*opresp.(*servicediscovery.GetOperationOutput).Operation.Targets["NAMESPACE"])
+	return nil
+}
+
+func resourceAwsServiceDiscoveryPublicDnsNamespaceRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).sdconn
+
+	input := &servicediscovery.GetNamespaceInput{
+		Id: aws.String(d.Id()),
+	}
+
+	resp, err := conn.GetNamespace(input)
+	if err != nil {
+		if isAWSErr(err, servicediscovery.ErrCodeNamespaceNotFound, "") {
+			d.SetId("")
+			return nil
+		}
+		return err
+	}
+
+	d.Set("description", resp.Namespace.Description)
+	d.Set("arn", resp.Namespace.Arn)
+	if resp.Namespace.Properties != nil {
+		d.Set("hosted_zone", resp.Namespace.Properties.DnsProperties.HostedZoneId)
```

Comment:
I wonder if Amazon has plans for expanding `Namespace.Properties` or `Namespace.Properties.DnsProperties`.

It's no big deal though as we can always deprecate this 1st-level field in favour of its nested version if it becomes necessary.